### PR TITLE
FDW changes for JdbcTest

### PR DESCRIPTION
**Pull request by hpandeycodeit**
This PR fixes/modifies JdbcTest to make it work with FDW (Readbale Tables).  

Following Tests are passing for FDW with this change: 

```
JdbcTest.jdbcColumnProjection...OK
JdbcTest.jdbcColumns...OK
JdbcTest.jdbcNamedQuery...OK
JdbcTest.jdbcReadableTableNoBatch...OK
JdbcTest.multipleFragmentsTables...OK
JdbcTest.readServerConfig...OK
JdbcTest.readViewSessionParams...OK
JdbcTest.singleFragmentTable...OK
```

### CODE REVIEW

File: automation/src/main/java/org/greenplum/pxf/automation/structures/tables/utils/TableFactory.java [hunk 1]
```diff
@@ -496,7 +496,7 @@ private static ExternalTable getPxfJdbcReadableTable(String name,
         if (user != null) {
             userParameters.add("USER=" + user);
         }
-        if (server != null) {
+        if (server != null && !FDWUtils.useFDW ) {
```
Comment (denalex) [inline automation/src/main/java/org/greenplum/pxf/automation/structures/tables/utils/TableFactory.java:499]: why is this needed ? `ForeignTable` knows how to parse this when creating `CREATE TABLE ..` SQL statement

File: automation/src/test/java/org/greenplum/pxf/automation/features/jdbc/JdbcTest.java [hunk 2]
```diff
@@ -156,6 +160,18 @@ private void prepareTypesData() throws Exception {
...
+        // $PXF_BASE/servers/database/reports.sql
+        if(FDWUtils.useFDW) {
+            gpdb.runQuery("DROP TABLE IF EXISTS FDW_JDBC_RESOURCE", true, false);
+            String fdw_resource_table = "CREATE TABLE FDW_JDBC_RESOURCE AS SELECT gpdb_dept.name, count(*), max(gpdb_emp.salary) " +
+                    "FROM gpdb_dept JOIN gpdb_emp " +
+                    "ON gpdb_dept.id = gpdb_emp.dept_id " +
+                    "WHERE gpdb_dept.id < 10 " +
+                    "GROUP BY gpdb_dept.name DISTRIBUTED BY (name);";
+            gpdb.runQuery(fdw_resource_table);
+        }
```
Comment (denalex) [inline automation/src/test/java/org/greenplum/pxf/automation/features/jdbc/JdbcTest.java:174]: I am not sure this is needed, see comment below

File: automation/src/test/java/org/greenplum/pxf/automation/features/jdbc/JdbcTest.java [hunk 3]
```diff
@@ -396,21 +413,31 @@ public void readServerConfig() throws Exception {
         runTincTest("pxf.features.jdbc.server_config.runTest");
     }
 
+    // For external table, this test reads the data (client_min_messages and default_statistics_target)
+    // from the jdbs-site.xml and update the pg_settings.
+    // For FDW, still needs to figure out how and where to set these params to update the pg_settings.
```
Comment (denalex) [inline automation/src/test/java/org/greenplum/pxf/automation/features/jdbc/JdbcTest.java:418]: why would this be any different for FDW ?

File: automation/src/main/java/org/greenplum/pxf/automation/components/gpdb/Gpdb.java [hunk 4]
```diff
@@ -196,8 +196,21 @@ private void createForeignServers(boolean ignoreFail) throws Exception {
...
-			runQuery(String.format("CREATE SERVER %s %s FOREIGN DATA WRAPPER %s OPTIONS(config '%s')",
-					option, foreignServerName, fdwName, pxfServerName), ignoreFail, false);
+			String jdbc_server_string = "";
+			// We need extra information for the server default_jdbc for the tests like jdbc_driver and db_url
+			// In case of external table, this information is read directly from the jdbc-site.xml
+			if(FDWUtils.useFDW && foreignServerName.equals("default_jdbc")) {
+				// create jdbc connection string
+				jdbc_server_string = ", jdbc_driver 'org.postgresql.Driver'," +
+						"db_url 'jdbc:postgresql://localhost:7000/pxfautomation'";
+				runQuery(String.format("CREATE SERVER %s %s FOREIGN DATA WRAPPER %s OPTIONS(config '%s' " + jdbc_server_string +")",
```
Comment (denalex) [inline automation/src/main/java/org/greenplum/pxf/automation/components/gpdb/Gpdb.java:206]: why do we need this ? While conceptually this is the right thing to do to associate JDBC driver and url with the FOREIGN SERVER, in external table these are table options, so for simplicity they will be auto-translated into FOREIGN TABLE options and things should work without this particular change.

File: automation/src/main/java/org/greenplum/pxf/automation/structures/tables/utils/TableFactory.java (hunk 1 above)
Comment (hpandeycodeit) [reply automation/src/main/java/org/greenplum/pxf/automation/structures/tables/utils/TableFactory.java:499]: In this case the FDW table gets created with two server options: 

For eg: 

```
CREATE FOREIGN TABLE pxf_jdbc_read_named_query_partitioned (name  text, count int, max  int) SERVER default_jdbc OPTIONS (resource 'query:report',PARTITION_BY 'count:int',RANGE '1:5',INTERVAL '1',SERVER 'database')

 pxfautomation=# \d+ pxf_jdbc_read_named_query
                            Foreign table "public.pxf_jdbc_read_named_query"
 Column |  Type   | Collation | Nullable | Default | FDW options | Storage  | Stats target | Description 
--------+---------+-----------+----------+---------+-------------+----------+--------------+-------------
 name   | text    |           |          |         |             | extended |              | 
 count  | integer |           |          |         |             | plain    |              | 
 max    | integer |           |          |         |             | plain    |              | 
Server: default_jdbc
FDW options: (resource 'query:report', server 'database') 
```

And then it fails with this error: 

```
< ERROR:  PXF server error : Invalid server name 'default_jdbc,database'

```

File: automation/src/main/java/org/greenplum/pxf/automation/structures/tables/utils/TableFactory.java [hunk 5]
```diff
@@ -496,8 +496,8 @@ private static ExternalTable getPxfJdbcReadableTable(String name,
         if (user != null) {
             userParameters.add("USER=" + user);
         }
-        if (server != null) {
-            userParameters.add("SERVER=" + server);
+        if (server != null ) {
```
Comment (ashuka24) [inline automation/src/main/java/org/greenplum/pxf/automation/structures/tables/utils/TableFactory.java:499]: nit: 
```suggestion
        if (server != null) {

File: automation/src/main/java/org/greenplum/pxf/automation/structures/tables/pxf/ForeignTable.java [hunk 6]
```diff
@@ -39,7 +39,7 @@ protected String createServer() {
         String[] serverParameters = StringUtils.defaultIfBlank(getServer(), "default").split("=");
         // getServer() might return a string "server=<..>", strip the prefix
         int index = serverParameters.length > 1 ? 1 : 0;
-        return String.format(" SERVER %s_%s", serverParameters[index], getProtocol());
+        return String.format(" SERVER %s_%s", serverParameters[index].replace("-","_"), getProtocol());
```
Comment (ashuka24) [inline automation/src/main/java/org/greenplum/pxf/automation/structures/tables/pxf/ForeignTable.java:42]: this is due to FDW not accepting `-` in server names correct?

Comment (hpandeycodeit) [reply automation/src/main/java/org/greenplum/pxf/automation/structures/tables/pxf/ForeignTable.java:42]: That's correct!

**Comment by hpandeycodeit** [comment]
As per discussion with @denalex , The logic ( introduced in the first commit) is changed and is more simpler now.

File: automation/src/main/java/org/greenplum/pxf/automation/components/gpdb/Gpdb.java (hunk 4 above)
Comment (hpandeycodeit) [reply automation/src/main/java/org/greenplum/pxf/automation/components/gpdb/Gpdb.java]: Resolving this and above comments as per our discussion.

File: automation/src/main/java/org/greenplum/pxf/automation/components/gpdb/Gpdb.java [hunk 7]
```diff
@@ -173,8 +173,11 @@ private void createForeignServers(boolean ignoreFail) throws Exception {
 		List<String> servers = Lists.newArrayList(
 		"default_hdfs",
 		"default_hive",
+		"db_hive_jdbc", //Needed for JdbcHiveTest
```
Comment (ashuka24) [inline automation/src/main/java/org/greenplum/pxf/automation/components/gpdb/Gpdb.java:176]: nit here and below as well
```suggestion
		"db_hive_jdbc", // Needed for JdbcHiveTest

File: automation/src/test/java/org/greenplum/pxf/automation/features/jdbc/JdbcTest.java [hunk 8]
```diff
@@ -401,16 +404,22 @@ public void readViewSessionParams() throws Exception {
...
 
+    @FailsWithFDW
     @Test(groups = {"features", "gpdb", "security", "jdbc"})
     public void jdbcWritableTable() throws Exception {
         runTincTest("pxf.features.jdbc.writable.runTest");
     }
 
+    @FailsWithFDW
+    // All the Writable Tests are failing with this Error:
+    //< ERROR:  PXF server error : class java.io.DataInputStream cannot be cast to class
```
Comment (ashuka24) [inline automation/src/test/java/org/greenplum/pxf/automation/features/jdbc/JdbcTest.java:415]:
```suggestion
    // ERROR:  PXF server error : class java.io.DataInputStream cannot be cast to class

**Review by ashuka24 — APPROVED** [review]

File: automation/src/test/java/org/greenplum/pxf/automation/features/jdbc/JdbcTest.java [hunk 9]
```diff
@@ -401,16 +404,22 @@ public void readViewSessionParams() throws Exception {
         runTincTest("pxf.features.jdbc.session_params.runTest");
     }
 
+    @FailsWithFDW
```
Comment (yimingli-vmware) [inline automation/src/test/java/org/greenplum/pxf/automation/features/jdbc/JdbcTest.java:407]: the `JdbcTest` is annotated with `WorkWithFDW`. Does `FailsWithFDW` override `WorkWithFDW`?

File: automation/src/test/java/org/greenplum/pxf/automation/features/jdbc/JdbcTest.java [hunk 10]
```diff
@@ -401,16 +404,22 @@ public void readViewSessionParams() throws Exception {
...
     @Test(groups = {"features", "gpdb", "security", "jdbc"})
     public void jdbcWritableTable() throws Exception {
         runTincTest("pxf.features.jdbc.writable.runTest");
     }
 
+    @FailsWithFDW
+    // All the Writable Tests are failing with this Error:
```
Comment (yimingli-vmware) [inline automation/src/test/java/org/greenplum/pxf/automation/features/jdbc/JdbcTest.java:414]: Since `jdbcWritableTable` is the first writable table test, will it be more clearer if we put this comment under that test?

**Review by yimingli-vmware — APPROVED** [review]

**Review by denalex — APPROVED** [review]
